From 7bbe558e91e04ca402a7eb33ad601fab649b202a Mon Sep 17 00:00:00 2001
From: Devi Priya <quic_devipriy@quicinc.com>
Date: Wed, 5 Jul 2023 15:19:46 +0530
Subject: [PATCH] usb: dwc3-qcom: fix clock handling in dwc3_qcom_remove

In dwc3_qcom_remove, dwc3_qcom_suspend is being called from
pm_runtime_allow function, after switching off the clocks.
In dwc3_qcom_suspend, qscratch region is being accessed.
This results in a crash due to access of un-clocked region.
So, moved the clock off logic after all the PM calls.

Also defer from calling clk_disable_unprepare if is_suspended
is set, since clocks are disabled during dwc3_qcom_suspend.

Change-Id: I75f624f951a872e68c593e588e7d8ee92c52f970
Signed-off-by: Balaji Prakash J <bjagadee@codeaurora.org>
(cherry picked from commit 17b5954738cd2a08843a89cee03529c5ce50f452)
Signed-off-by: Devi Priya <quic_devipriy@quicinc.com>
---
 drivers/usb/dwc3/dwc3-qcom.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-qcom.c b/drivers/usb/dwc3/dwc3-qcom.c
index 7ba76e97c440..1c44a8bfefc2 100644
--- a/drivers/usb/dwc3/dwc3-qcom.c
+++ b/drivers/usb/dwc3/dwc3-qcom.c
@@ -990,12 +990,6 @@ static int dwc3_qcom_remove(struct platform_device *pdev)
 	device_remove_software_node(&qcom->dwc3->dev);
 	of_platform_depopulate(dev);
 
-	for (i = qcom->num_clocks - 1; i >= 0; i--) {
-		clk_disable_unprepare(qcom->clks[i]);
-		clk_put(qcom->clks[i]);
-	}
-	qcom->num_clocks = 0;
-
 	dwc3_qcom_interconnect_exit(qcom);
 	reset_control_assert(qcom->resets);
 
@@ -1011,6 +1005,13 @@ static int dwc3_qcom_remove(struct platform_device *pdev)
 	pm_runtime_allow(dev);
 	pm_runtime_disable(dev);
 
+	for (i = qcom->num_clocks - 1; i >= 0; i--) {
+		if (!qcom->is_suspended)
+			clk_disable_unprepare(qcom->clks[i]);
+		clk_put(qcom->clks[i]);
+	}
+	qcom->num_clocks = 0;
+
 	return 0;
 }
 
-- 
2.34.1

